File: clojure/binary-search-tree/src/binary_search_tree.clj
(ns binary-search-tree)

(defn value [tree-node]
  (:value tree-node))

(defn singleton [value]
  {:value value :left nil :right nil})

(defn insert [value root]
  (cond
    (nil? root) (singleton value)
    (> value (:value root)) (update root :right (partial insert value))
    :else (update root :left (partial insert value))))

(defn left [tree-node]
  (:left tree-node))

(defn right [tree-node] 
  (:right tree-node))

(defn to-list [root]
  (if (nil? root) []
    (concat (to-list (:left root)) [(:value root)] (to-list (:right root)))))

(defn from-list [l] 
  (reduce #(insert %2 %1) nil l))

File: ocaml/binary-search-tree/binary_search_tree.ml
open Base

type bst = Nil | Node of (bst * int * bst)

let empty: bst = Nil

let value = function
    | Nil -> Error "tree is Nil"
    | Node (_, v, _) -> Ok v

let left = function
    | Nil -> Error "tree is Nil"
    | Node (l, _, _) -> Ok l

let right = function
    | Nil -> Error "tree is Nil"
    | Node (_, _, r) -> Ok r

let rec insert v = function
    | Nil -> Node (Nil, v, Nil)
    | Node (l, v2, r) when v2 >= v -> Node(insert v l, v2, r)
    | Node (l, v2, r) -> Node(l, v2, insert v r)

let rec to_list = function
    | Nil -> []
    | Node(l , v, r) -> (to_list l) @ v @ (to_list r)
    
    

